module/zfs/dbuf.c

```diff
@@ -1027,6 +1027,20 @@ dbuf_read_impl(dmu_buf_impl_t *db, zio_t *zio, uint32_t flags)
 		 */
 		int bonuslen = MIN(dn->dn_bonuslen, dn->dn_phys->dn_bonuslen);
 		int max_bonuslen = DN_SLOTS_TO_BONUSLEN(dn->dn_num_slots);
+		arc_buf_t *dn_buf = (dn->dn_dbuf != NULL) ?
+		    dn->dn_dbuf->db_buf : NULL;
+
+		if (dn_buf != NULL && arc_is_encrypted(dn_buf) &&
+		    DMU_OT_IS_ENCRYPTED(dn->dn_bonustype) &&
+		    !(flags & DB_RF_NO_DECRYPT)) {
+			err = arc_untransform(dn_buf, dn->dn_objset->os_spa,
+			    dmu_objset_id(dn->dn_objset), B_TRUE);
```

Comment:
This is something I discussed with you a while back, but I might not have explained it very well. I was unable to find a good way to lock around this code because `dn_buf` might be shared with many dnodes and there is no synchronization that currently exists to coordinate access to the underlying arc buffer. Technically this code can race, but all threads will want the same end result, all bonus buffers decrypted. I don't like this solution, but I was hard pressed to find anything else that could work.

Can you think of a better way to handle this? I looked back at the corresponding code in `arc_buf_fill()` and I see that I made a small mistake in accounting for this racy behavior. I'm afraid other developers that touch this code might make the same kind of mistake because its not great practice.